### ISSUE TITLE
Handling nil/false attributes and attr defaults consistently

### ISSUE THREAD

**Issue by thiagomajesk** [phoenixframework/phoenix_live_view#3602]
Typically `nil` and `false` will omit the attribute as per the documentation:
https://hexdocs.pm/phoenix_live_view/Phoenix.Component.html#sigil_H/2-syntax

> false or nil - if a value is false or nil, the attribute is omitted. Some attributes may be rendered with an empty value, for optimization purposes, if it has the same effect as omitting [...]

So, the following:

```html
<span class={nil}>Hello World</span>
<span class={false}>Hello World</span>
```

Will output: `<span class="">Hello World</span>`

And the following:

```html
<span data-class={nil}>Hello World</span>
<span data-class={false}>Hello World</span>
```

Will output: `<span>Hello World</span>`

It gets especially weird because this is not how it's handled if you have a function component with global attributes:

````elixir
attr :rest, :global
slot :inner_block

defp span(assigns), do: ~H"<span {@rest}>{render_slot(@inner_block)}</span>"
````

The following:

```html
<.span class={nil}>Hello World</.span>
<.span class={false}>Hello World</.span>
<.span data-class={nil}>Hello World</.span>
<.span data-class={false}>Hello World</.span>
```

Will output: `<span data-phx-id="m3-phx-GBS__HJfKXiKyBHD">Hello World</span>`

I don't know exactly what optimizations the doc is referring to, but to me, it already feels too implicit/ hard to track.

--- 

And it gets a little trickier because you can bypass the compiler warning with something like the following:

```elixir
attr :class, :string, default: "foo"
slot :inner_block

defp span(assigns), do: ~H"<span class={@class}>{render_slot(@inner_block)}</span>"
```

Invoking it like this ` <.span class={if nil, do: nil}>Hello World</.span>` also won't pick up the default value of the attr.

So, "nillish" values are handled in 3 different ways. Given that conditional logic with attrs is pretty common, should we try to make this more consistent for better DX?

- Always empty attributes with `nil` unless it's a boolean attribute!?
- Prevent compiler type checking attr when `nil` (unless attr is required)!?
- Empty attr assign / use default value when `nil` is given to attr

**Comment by josevalim** [comment]
There are two separate discussions here:

1. Dealing of `class` in HTML attributes for optimization purposes
2. Dealing of any attribute when given nil

When you write it like this:

    <.span class={if nil, do: nil}>Hello World</.span>

It is completely unrelated to `1` and with it being called `class`. If a value is given, then the default is never used. That's consistent and clear behaviour.

Meanwhile, the optimization itself only affects the HTML sent over the wire, and you should not see any difference in behaviour. So I think the issue here is the documentation not being precise enough and making you think the optimization is more general or applies to more scenarios than it does.

**Comment by thiagomajesk** [comment]
> It is completely unrelated to 1 and with it being called class. If a value is given to a component, then the default is never used, regardless of the attribute. That's consistent and clear behaviour for all attributes.

I think you are missing the point, this is not even related to the optimization at all, it's related to the fact that special values (`nil` and `false`) are handled differently between HTML and function components for all attributes. 

You might claim that this is consistent behavior, but it's extremely annoying/ confusing that it works that way. Also, I don't think that the doc changes on https://github.com/phoenixframework/phoenix_live_view/commit/f5651fba2cfc35a6688584f2ef5db01bee681c9a is related to the issue at all, since this issue I'm pointing out will happen for other attributes (like `data-class` in my example). 

I think most people coming to LV from other frameworks would fully expect to be able to refactor HTML code into components and expect both to have the same behavior when dealing with attributes, and that's just not how it works today.

**Comment by josevalim** [comment]
How does it behave differently? If your component abstracts HTML and then forward its attributes, with `attr={@attr}`, then it will render the same. Can you tell me how they are different without using the `class` attribute?

PS: please keep the discussion focused. Saying “I think most people coming to LV from other frameworks would fully expect X” does not help the discussion, because without actually running a poll and asking people’s opinion, your guess is as good as mine.

**Comment by thiagomajesk** [comment]
My example with the `data-class` attribute is one. On HEEx , if you pass either `nil` or `false` it won't render the attribute, but with function components, `nil` is considered a valid value, even for attributes that are not required. 

We don't need to run a pool for this, the fact that HEEx works in a specific way and the semantics don't translate to function components is what makes me think that this will confuse people. It`s not a so far-fetched conclusion.

**Comment by thiagomajesk** [comment]
I just updated my previous comment with a better example, in other words, I think what I'm trying to say is that:

`<span other={nil}>Hello World</span>` and `<.span other={nil}>Hello World</.span>` should work the same, but that's only true if you use `rest` (global attributes). If you have specified the `attr` you get `nil` and the `default` option only works if you omit the attribute completely (in HTML both will result in the attribute being omitted), then you need additional logic to specify a default value. Those things are interconnected to how you interact with attributes in HEEx vs function components.

**Comment by SteffenDE** [comment]
@thiagomajesk this is definitely not something that's going to change, as it would be a breaking change. I also don't think that it should change. Function components are more like calling a function that can return more HTML (HEEx), than actual HTML tags. Concerning "nil should use the default", I don't think that this is reasonable. It's similar to `Map.get(%{foo: nil}, :foo, :default)` where you also expect to only get the default value if it was not given at all. Any changes to the documentation to make this clearer are very welcome though, as always.

**Comment by thiagomajesk** [comment]
@SteffenDE It might not be such a breaking change because if you don't have `default` specified for an `attr`, the value would still be `nil`, the issue is when you have `default` specified and it only works if you omit the value in the HTML. TBH, I don't think is such an unreasonable expectation if you look closer. 

Not having default being respected, makes composing `attr` kinda difficult, because you have to account for when the client omitted the attribute plus `nil` (like if some conditional logic was used), for instance:

```elixir
attr :value, :integer, default: 0
attr :max_value, :integer, default: 100

defp progress(assigns) do
  value = assigns.value || 0
  percentage = [...]
  assigns = assign(assigns, :percentage, percentage)  
  ~H"""
  <div>{@percentage} %</div>
  """
end
```

Respecting `default` here doesn't introduce a breaking change and for the cases where `default` is not there, you can assume, `default:  nil`.

**Comment by josevalim** [comment]
Thanks for removing the `class`/optimization from the discussion, that makes it much easier to track. There are a few things going on here, so let's break it down.

As the compiler warnings points out, if you declare an attribute `attr :other, :integer, default: 0`, `nil` is not a valid value to be given as input. The fact you can hide it with a conditional is a limitation of how we perform the checking. However, when we integrate LV with the type system in the future, you would get a warning for the conditional example too:

    <.span other={if nil, do: nil} data-other={false}>Hello World</.span>

The good news about `nil` being an invalid input today is that we can choose to support it, if we want to.

### HTML equivalence

Part of the discussion has been around HTML equivalence. I'd argue that, if your goal is to be equivalent to HTML, then you should declare it as:

    attr :value, :any

That's because an HTML attribute can accept a value of `:any` type (as long as it implements the HTML Safe protocol). In that sense, setting an attribute to `:integer` is effectively a subset of the HTML API (a subset that only accepts integers, and not even `nil`).

Using `attr :value, :any, default: :something` is also a good place to help us understand the behaviour of `value={nil}`. Today, if you pass `value={nil}` to an attribute of `:any` type, `@value` will be set to `nil`, and ignore the default. As @SteffenDE pointed out, this is consistent with how maps and keyword lists behave in Elixir, where **`value: nil` is not the same as not passing `value:`**. Changing this behaviour would be backwards incompatible and inconsistent.

### Should we support nils in `:integer` and other types?

Now that we defined that `value={nil}` works as maps in Elixir and is consistent with HTML when the type is `:any`, the question is: should we add support for `nil` in `:integer`?

We certainly should not do it _by default_. If we suddenly accept nils everywhere (effectively making all types nullable), now we can start hiding real bugs in the code. Maybe someone declares:

    attr :value, :integer, default: 123

And if we suddenly accept `nil`, now the component code will receive `nil`, which can lead to subtle bugs. We could also have bugs for the person calling the component. If `value={user.something}` returns nil, is it because of a bug in the code or because you want to reset it to the default value? The direction most languages and frameworks have evolved in the last two decades is to make nil types opt-in, instead of opt-out. You are right that existing attributes without an explicit default value already needs to deal with `nil`, but there is a big difference between: "I handle `nil` as part of my _internal_ defaults" vs "I handle `nil` as part of my API".

Alternatively, you could argue that we could allow `nil` for the `value` attribute of `:integer` type and that would automatically set it as a default. Therefore the component author does not have to deal with `nil`. However, that would be inconsistent with how the `:any` type works, as it does not reset it to a default value.

Therefore, at best, we could add a `nullable: true` (or `allow_nils: true`), where you could explicitly say that an attribute can be set to nil (and that would also be supported by the type system in the future). However, I'd postpone this for now as the type system would allow us to do all of this in much more expressive ways. If you want nils to be given, using `:any` is fine today.

**Comment by thiagomajesk** [comment]
@josevalim Just a small correction around the HTML equivalence part, I'm limiting my example to the fact that when you specify `nil` in HEEx you get an empty value (I didn't even consider `any` as you mentioned). I think there's a point to be made here that if you don't want the compile complaining to you about the `attr` you should ignore the type definition with something like `any` (which makes sense and is easy to explain to users). |

> Changing this behaviour would be backwards incompatible and inconsistent with Elixir.

I understand the analogy with maps here, but I think this is not what the average user would expect while dealing with attributes. That's why I tried to draw a parallel with HEEx, which I think is closer to the mental model than maps. I hope that clarifies it 👍.

 > We certainly should not do it by default. If we suddenly accept nils everywhere (effectively making all types nullable), now we can start hiding real bugs in the code
 
 > And if we suddenly accept nil, now the component code will receive nil, which can lead to subtle bugs
 
Here's my problem with this reasoning, we already have `required: true`, so why should the compiler even care when the value is optional? In this example: `attr :value, :integer, default: 123`, I would expect that calling `value={nil}`, would result in `123`, not `nil`; so it's not about making things nullable (they already are when you do conditionals because the compiler doesn`t check), the other way around. It's not about making things nilable, it's actually using the already specified `default` value in `attr`. And then if you expect `nil` you would declare it as attr :value, :integer` or `attr :value, :integer, default: nil`. 

I understand the suggestion of using something like `nullable: true` in a scenario where the compiler can also check the conditionals, but the behavior I'm looking for is actually `nullable: false` + `default: 123` resulting in `123` when attr is `nil`. 

Today is like every attr is already `nullable: true` by default, and I think it might be harder to explain that  `nullable: false` will use actually use the `default` value in attr than just having required/ optional attributes. 

I think it might be easier to explain that `required: true` means that the compiler will check the `attr` is present and `required: false` (the default) won't. Additionally, `default: 123` specifies the default value, and `default: nil` allows for `nil` values.

**Comment by thiagomajesk** [comment]
Sorry for the amount of edits, I have a new keyboard that is driving me crazy...
Just to clarify something about the compiler check, IMO, if the compiler checks something like this: `<.span other={if nil, do: nil} data-other={false}>Hello World</.span>` from my previous example and complains that `other` should be an integer, it feels like its behaving like `required: true`, forcing me to pass a value. This means again, that the caller now has to worry about using a default value instead of using what we already have in `default`.

**Comment by josevalim** [comment]
> but I think this is not what the average user would expect while dealing with attributes

Once again, nor me or you knows what the average user expects without actually asking them and running a poll. You can totally say it is your preference though, that's fine. :) Otherwise these generalizations are not productive because I could just claim the opposite. 

> In this example: `attr :value, :integer, default: 123`, I would expect that calling value={nil}, would result in 123, not nil

I explained why this is not a good idea above. To recap:

1. That's not how it behaves with `:any` (and we can't change it, it would be backwards incompatible)

2. That's not how it behaves with maps

3. It can accidentally hide bugs where the user is accidentally passing nil. Nullable types are opt-in in most recent languages. It is opt-in in Elixir's type system. It is opt-in in TypeScript. If we want to allow nullable here, it should 100% be opt-in

> Just to clarify something about the compiler check, IMO, if the compiler checks something like this: <.span other={if nil, do: nil} data-other={false}>Hello World</.span> from my previous example and complains that other should be an integer, it feels like its behaving like required: true, forcing me to pass a value. 

You are confusing the presence/absence of a value with the value being nullable or not. Being required means `other={...}` has to be given, it does not say if the value should be `nil`, `:integer` or something else. Requiring the value to be an integer (not nil) is different from required.

We cannot show this in Elixir type system yet but I setup some TypeScript code for you to see the differences:

https://www.typescriptlang.org/play/?#code/FAFwngDgpgBA8hEBLA9gOwIYBsByKQ4CuWWMAvDAN4wAeA-AFwwAUahAtgEZQBOMAPjADOIHkjQBzAJQwAvgG5QkWAmTpsREuSq1GLNl14Dho8RONsSMhUugwASlACOhJDygATPAWKkK1GiZWDm4+QRExSWtFcDtHFzdPTT8dQP0Qo3DTSQtfaOBgLCgQGCh2RDAABiZVVExcfGTtahsikoA3bCQPavhEOo1G32baJgBGACYAZjlFNphxTqxu3tr1Bp8tf1GYS1IbQuLS8vAxmv71pu3Wo6Xus761eqvU8enZw5LFro8Htefhts0nsPp9jhUJkx4q53F4hlsdDcOj9IQ5nDCkvCUgE3jMkQs0HcPKjoYk4ZtsTsQQd5mUKlMoeiyS8WnNbj8GWiErCWTtJni2V9CRzGdykoDXrthjYgA

In particular, look at this example:

```ts
type OptionalNotNull = { x?: (number | string) };
let invalid0: OptionalNotNull = { x: null };
```

It reports a type violation because, the fact it is optional, does not allow you to set it null. `nullable` is a different property than optional.

**Comment by josevalim** [comment]
Btw, the best behaviour to achieve what you want today is to set it to `:any` and then do `@value || 123`. If we add `nullable: true`, then you can declare it as `attr :value, :integer, nullable: true` and do `@value || 123`.

**Comment by thiagomajesk** [comment]
> Once again, nor me or you knows what the average user expects without actually asking them and running a poll. You can totally say it is your preference though, that's fine. :) Otherwise these generalizations are not productive because I could just claim the opposite.

I honestly don't know what else I can do besides prefacing what I'm saying with _"I think"_... However, I have to mention that I've seen multiple other people starting with LiveView getting confused by this behavior. So please, don't dismiss it as a generalization. There's always a component of preference in good DX that you can't capture with a binary survey.

> You are confusing the presence/absence of a field with the value being nullable or not. Being required means other={...} has to be given, but the type is the one that controls which values are valid once given. :integer does not accept nils regardless if required or optional.

That's a fair point you got there; I haven't touched a statically typed language in at least 7 years, so my type theory knowledge is a bit fuzzy, thanks for bearing with me that's an important distinction where we were diverging. I guess the surprise comes mainly from the fact that today you can cheat out of that by using a conditional (which is really common in complex component code). 

That being said, it seems that `default` is not that useful since it only covers the use case of the absence of the attribute. So, do you have any other solutions in mind so we don't have to always specify `default` + do the manual check in the component body?

It seems that `nullable` would be a good complement to the declarative attributes, but it doesn't solve the 'default' issue yet; what do you think about having a `coalesce: true` option that would cover both cases? (effectively, adding the value from `default` to the assigns if the value is `nil`)

**Comment by josevalim** [comment]
> However, I have to mention that I've seen multiple other people starting with LiveView getting confused by this behavior. 

Yes, but there could also be that 80% of people are happy with the current behaviour, and they naturally just don't say anything. So once again, let's avoid generalizations, unless we have data to back it up.

> That's a fair point you got there; I haven't touched a statically typed language in at least 7 years, so my type theory knowledge is a bit fuzzy, thanks for bearing with me that's an important distinction where we were diverging.

FWIW, you have this same distinction in typespecs and, as @SteffenDE pointed out, it also happens when doing things like `Map.get(%{foo: nil}, :foo, 123)`.

> what do you think about having a coalesce: true option that would cover both cases? (effectively, adding the value from default to the assigns if the value is nil)

I don't see reasons to add this option. As I said before, this is not something common with Elixir APIs and we don't want to promote passing `nil`s everywhere. So if you really want this behaviour, you can achieve this with `default: nil` and then doing `@value || default`.

I am bowing out of this discussion for now, as I am mostly rehashing past arguments. Thanks.

**Comment by josevalim** [comment]
> However, I have to mention that I've seen multiple other people starting with LiveView getting confused by this behavior. 

Yes, but there could also be that 80% of people are happy with the current behaviour, and they naturally just don't say anything. So once again, let's avoid generalizations. Preferences are fine, but we tend to naturally stick with the preference of the core team, unless we have data or concrete evidence that says otherwise. :)

> That's a fair point you got there; I haven't touched a statically typed language in at least 7 years, so my type theory knowledge is a bit fuzzy, thanks for bearing with me that's an important distinction where we were diverging.

FWIW, you have this same distinction in typespecs and, as @SteffenDE pointed out, it also happens when doing things like `Map.get(%{foo: nil}, :foo, 123)` (there is no "coalescing").

> what do you think about having a coalesce: true option that would cover both cases? (effectively, adding the value from default to the assigns if the value is nil)

I don't see reasons to add this option. As I said before, this is not something common within Elixir APIs and we don't want to promote passing `nil`s everywhere. So if you really want this behaviour, you can achieve this with `:any` and then doing `@value || default`.

I am bowing out of this discussion for now, as I am mostly rehashing past arguments. Thanks.
